### PR TITLE
fix: fix handling of intersection types for lambda parameters

### PR DESCRIPTION
**Pull request by quentinLeDilavrec**
#3520

### CODE REVIEW

File: src/main/java/spoon/support/reflect/code/CtLambdaImpl.java [hunk 1]
```diff
@@ -101,7 +105,32 @@ public String getSimpleName() {
 			//it can be null in noclasspath mode, so we do not know which method is called, by lambda
 			return null;
 		}
-		CtType<T> lambdaType = lambdaTypeRef.getTypeDeclaration();
+		CtMethod<?> lambdaExecutableMethod = null;
+		if (!(lambdaTypeRef instanceof CtIntersectionTypeReference)) {
+			return getOverriddenAux(lambdaTypeRef);
+		} else {
+			CtElement parent = lambdaTypeRef.getParent();
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/code/CtLambdaImpl.java:112]: refactor the else block in `getOverriddenMethodForIntersectionType`?

File: src/main/java/spoon/support/reflect/code/CtLambdaImpl.java [hunk 2]
```diff
@@ -101,7 +105,32 @@ public String getSimpleName() {
...
+				parentTypeReference = ((CtAssignment) parent).getAssigned().getType();
+			}
+			for (CtTypeReference<?> ctTypeReference : ((CtIntersectionTypeReference<?>) lambdaTypeRef).getBounds()) {
+				CtMethod<?> tmp = getOverriddenAux(ctTypeReference);
+				if (tmp != null && (lambdaExecutableMethod == null || ctTypeReference.equals(parentTypeReference))) {
+					lambdaExecutableMethod = tmp;
+				}
+			}
+		}
+		if (lambdaExecutableMethod == null) {
```
Comment (monperrus) [inline src/main/java/spoon/support/reflect/code/CtLambdaImpl.java:126]: do we need to move this check here? could we keep it at the previous place?

File: src/test/java/spoon/test/lambda/LambdaTest.java [hunk 3]
```diff
@@ -458,6 +458,16 @@ public void testLambdaWithGenericExtendingMultipleInterfaces() {
 		assertIsWellPrinted("( elt) -> elt.test()", lambda);
 	}
 
+	@Test
+	public void testCastLambdaWithIntersection() {
+		final CtLambda<?> lambda1 = intersection.getElements(new TypeFilter<CtLambda<?>>(CtLambda.class)).get(1);
```
Comment (monperrus) [inline src/test/java/spoon/test/lambda/LambdaTest.java:463]: could you add a one line comment stating the test intention in natural language:

    // contract: intersection types on lambda parameters are supported

File: src/main/java/spoon/support/reflect/code/CtLambdaImpl.java (hunk 2 above)
Comment (quentinLeDilavrec) [reply src/main/java/spoon/support/reflect/code/CtLambdaImpl.java:126]: It follows the standard compiler behavior.
That is, at least one interface should have one method,
thus, some of these interfaces might not even declare any methods.

But as a side note, I just match the type that will be taken by the lambda in its parent,
but I did not implement a real type intersection, thus I am less strict than the compiler.
A great improvement would be to implement this type intersection in spoon.
But it is too complicated for me :/ and I would not be useful for my use case.

Comment (monperrus) [reply src/main/java/spoon/support/reflect/code/CtLambdaImpl.java:112]: OK for me, thanks.

**Comment by monperrus** [comment]
Thanks. We have a few checkstyle problems.

**Comment by nharrand** [comment]
LGTM! @monperrus do you want to merge?

**Comment by monperrus** [comment]
Yes, thanks a lot @quentinLeDilavrec and @nharrand for the polishing commit!

**Pull request merged**